File: Samples/Objc/iOS/View Model/PlayerViewModel.m
//
//  PlayerViewModel.m
//  iOSObjc
//
//  Created by nice on 18/11/2019.
//  Copyright © 2019 nice. All rights reserved.
//

#import "PlayerViewModel.h"
#import <YouboraLib/YouboraLib.h>
#import <YouboraConfigUtils/YouboraConfigUtils-Swift.h>
#import <YouboraSpotXPlayerAdapter/YouboraSpotXPlayerAdapter-Swift.h>
#import <YouboraAVPlayerAdapter/YouboraAVPlayerAdapter.h>
#import "Constants.h"

@interface PlayerViewModel()

@property YBPlugin *plugin;

@end


@implementation PlayerViewModel

-(instancetype)initWithSegueIdentifier:(NSString*)segueIdentifier {
    self = [super init];
    
    if (self) {
        self.selectedOption = [Utils translateToSegue:segueIdentifier];
        
        YBOptions *options = [YouboraConfigManager getOptions];
        options.offline = NO;
        options.waitForMetadata = NO;
        
        options.adExpectedPattern = @{@"pre":[NSNumber numberWithInt:1]};
        
        self.plugin = [[YBPlugin alloc] initWithOptions:options];
    }
    
    return self;
}

-(void)setAdsAdapter:(SpotXInterstitialAdPlayer*)player andDelegate:(id<SpotXAdPlayerDelegate>)delegate {
    if (!player) { return; }
    
    YBPlayerAdapter *adapter = [[[YBSpotXPlayerAdAdapterManager alloc] initWithPlayer:player delegate:delegate] getAdapter];
    
    if(!player) { return; }
    
    self.plugin.adsAdapter = adapter;
}

-(void)setAdapter:(AVPlayer*)player {
    self.plugin.adapter = [[ YBAVPlayerAdapter alloc ] initWithPlayer:player];
}

-(SpotXAdRequest*)getAdRequest {
    SpotXAdRequest *request = [SpotXAdRequest new];
    [request setChannel:SPOT_X_TEST_CHANNEL];
    
    return request;
}

-(NSURL*)getVideoUrl {
    return [[NSURL alloc] initWithString:VIDEO_URL];
}

-(void)viewWillDisappear {
    [self.plugin removeAdsAdapter];
    [self.plugin removeAdapter];
}

@end

File: Samples/Swift/iOS/View Models/PlayerViewModel.swift
//
//  BasicPlayerViewModel.swift
//  iOSSwift
//
//  Created by nice on 04/11/2019.
//  Copyright © 2019 nice. All rights reserved.
//

import Foundation
import SpotX
import YouboraLib
import YouboraConfigUtils
import AVKit
import YouboraAVPlayerAdapter
import YouboraSpotXPlayerAdapter

class PlayerViewModel {
    public let selectedOption: MenuSegue
    private let plugin: YBPlugin
    
    var wrapper: YBAVPlayerAdapterSwiftWrapper?
    
    public var videoUrl: URL? { return URL(string: Constants.normalVideoUrl) }
    
    init(segueIdentifier: String?) {
        if let segueIdentifier = segueIdentifier,
            let segue = MenuSegue(rawValue: segueIdentifier) {
                self.selectedOption = segue
        } else {
            self.selectedOption = .unknown
        }
        
        let options = YouboraConfigManager.getOptions()
        
        options.offline = false
        options.waitForMetadata = false
        
        self.plugin = YBPlugin(options: options)
        
    }
    
    func getAdRequest() -> SpotXAdRequest {
        
            let request = SpotXAdRequest()
            
            request.setChannel(Constants.spotXTestChannel)
        
            return request
        }
    
    func setAdapter(player: AVPlayer?) {
        guard let nonOptionalPlayer = player else {
            return
        }
        
        self.plugin.fireInit()
        self.wrapper = YBAVPlayerAdapterSwiftWrapper(adapter: YBAVPlayerAdapter(player: nonOptionalPlayer), andPlugin: self.plugin)
    }
    
    func setAdsAdapter(player: SpotXInterstitialAdPlayer?, delegate: SpotXAdPlayerDelegate) {
        guard let player = player,
            let adapter = YBSpotXPlayerAdAdapterManager(player: player, delegate: delegate).getAdapter() as? YBPlayerAdapter<AnyObject> else {
            return
        }
        
        self.plugin.adsAdapter = adapter
    }
    
    func viewWillDisappear() {
        self.plugin.removeAdsAdapter()
        self.plugin.removeAdapter()
    }
}
